internal/util/observability/funccall.go

```diff
@@ -18,52 +18,65 @@ import (
 	"context"
 	"runtime"
 	"runtime/trace"
+	"sync"
 
 	"github.com/FerretDB/FerretDB/internal/util/resource"
 )
 
 // funcCall tracks function calls.
 type funcCall struct {
-	token  *resource.Token
-	region *trace.Region
+	cancelCtx context.CancelFunc
+	token     *resource.Token
+	region    *trace.Region
 }
 
-// FuncCall adds observability to a function call.
+// FuncCall adds observability to a function or method call.
+// It returns a new derived context and its cancel function.
+//
+// FuncCall should be called at the very beginning of the observable function,
+// and the returned cancel function should be called at exit.
+// The returned context should be passed down.
 //
-// It should be called at the very beginning of the function,
-// and returned function should be called at exit.
-// The returned function must not be passed or stored.
 // The only valid way to use FuncCall is:
 //
 //	func foo(ctx context.Context) {
-//	    defer FuncCall(ctx)()
-//	    // ...
+//		ctx, cancel := FuncCall(ctx)
+//		defer cancel()
+//
+//		bar(ctx)
+//		// ...
+//	}
 //
 // For the Go execution tracer, FuncCall creates a new region for the function call
 // and attaches it to the task in the context (or background task).
-func FuncCall(ctx context.Context) func() {
+func FuncCall(ctx context.Context) (context.Context, context.CancelFunc) {
+	ctx, cancelCtx := context.WithCancel(ctx)
+
 	fc := &funcCall{
-		token: resource.NewToken(),
+		cancelCtx: cancelCtx,
+		token:     resource.NewToken(),
 	}
 	resource.Track(fc, fc.token)
 
-	if trace.IsEnabled() {
-		pc := make([]uintptr, 1)
-		runtime.Callers(1, pc)
-		f, _ := runtime.CallersFrames(pc).Next()
-		funcName := f.Function
+	pc := make([]uintptr, 1)
+	runtime.Callers(1, pc)
+	f, _ := runtime.CallersFrames(pc).Next()
+	funcName := f.Function
 
+	if trace.IsEnabled() {
 		fc.region = trace.StartRegion(ctx, funcName)
 	}
 
-	return fc.leave
+	return ctx, sync.OnceFunc(fc.cancel)
 }
 
-// leave is called on function exit.
-func (fc *funcCall) leave() {
+// cancel is called on function exit.
+func (fc *funcCall) cancel() {
 	if fc.region != nil {
 		fc.region.End()
 	}
 
 	resource.Untrack(fc, fc.token)
+
+	fc.cancelCtx()
```

Comment:
The only idea I had - it would work if we don't cancel that context.

Reply:
The failure is interesting. Because I didn't expect this. We create a copy of original context, when function exits and cancel that context all operation should be done. So I didn't expect to see `context canceled` error.

Reply:
Actually reading the code, I understand what is happening. The `context canceled` came from integration test setup code to cleanup collection after test is completed.

See how observability is calling defer cancel?
https://github.com/FerretDB/FerretDB/blob/f79460f8b8a7973842c5dfe9005ca1a3cdb836dc/integration/setup/setup.go#L210-L211

See how `t.Cleanup()` is using canceled context because it was cancelled when that function exited. So it fails here.
https://github.com/FerretDB/FerretDB/blob/f79460f8b8a7973842c5dfe9005ca1a3cdb836dc/integration/setup/setup.go#L251-L258

What we can do is using a new context to call `t.Cleanup` or cancel the context in `t.Cleanup`. Which captures observability more accurately?

Reply:
We would have same issue in other places where `tb.Cleanup` is used within integration test, such as in compatSetup.

Reply:
Yes, also in functions with goroutines, and, for example, `db.QueryContext`. I haven't found the whole list yet, `tb.Cleanup` was my last finding yesterday. My biggest concern is if we have so many exceptions, is it worth it at all?

Reply:
I only spot two places, above and this https://github.com/FerretDB/FerretDB/blob/main/integration/setup/setup_compat.go#L158.

Where are other places?

Reply:
See the comments starting with `FuncCall's context can't be used`

Reply:
I see one place where it is used. It's where we start FerretDB with `l.Run(ctx)` right? https://github.com/FerretDB/FerretDB/blob/main/integration/setup/listener.go#L252

Yes this one is quite naturally not gonna work.

I have a question, how should observability work for go routine? Do we start new observability for go routine? Do we ignore observability for go routine? In any case it cannot share the same observability/ctx from the function that created go routine.

Reply:
I think it depends on how we want to capture observability on tools we use.

Reply:
> I have a question, how should observability work for go routine? Do we start new observability for go routine

To me, the question is what do we want to observe, how do we choose these places and why :)

And actually, I have an unpopular (at FerretDB) opinion here that with otel we should focus on request-based calls (however, in tests we do it differently). On the other hand, if we talk about pprof or metrics or execution tracing, these things don't have to be focused on requests, rather on interesting parts of the code.

Reply:
I think you have a really good point. It's true that what we want to focus differs based on what we are using.
Also true, the test is different. For example for pprof and metrics, things such as setup in test might be interesting (in my opinion)

Reply:
The current implementation detaches ctx and observability when go routine is around. Below are a few alternatives, code might not compile but just getting the idea across.

We could create observability for each go routine Can it even be useful for our tracking tool (maybe not?).
```
func setupListener(tb testtb.TB, ctx context.Context, logger *zap.Logger, opts *BackendOpts) string {
	tb.Helper()

	setupCtx, cancel := observability.FuncCall(ctx) // setupCtx is used for setting up a listener
	defer cancel()

	runDone := make(chan struct{})

	go func(ctx context.Context) {
		ctx, cancel := observability.FuncCall(ctx)
		defer cancel()

		defer close(runDone)

		l.Run(ctx)
	}(ctx)
```
We can ignore observability for go routine.
```
func setupListener(tb testtb.TB, ctx context.Context, logger *zap.Logger, opts *BackendOpts) string {
	tb.Helper()

	setupCtx, cancel := observability.FuncCall(ctx) // setupCtx is used for setting up a listener
	defer cancel()

	runDone := make(chan struct{})

	go func() {
		defer close(runDone)

		l.Run(ctx)
	}()
```

Reply:
> To me, the question is what do we want to observe, how do we choose these places and why :)

You just answered your own questions